dvc/remote/base.py

```diff
@@ -541,15 +553,40 @@ def changed_cache(self, checksum):
         return self.changed_cache_file(checksum)
 
     def cache_exists(self, checksums):
-        # NOTE: The reason for such an odd logic is that most of the remotes
-        # take much shorter time to just retrieve everything they have under
-        # a certain prefix(e.g. s3, gs, ssh, hdfs). Other remotes that can
-        # check if particular file exists much quicker, use their own
-        # implementation of cache_exists(see http, local).
-        #
-        # Result of all() might be way too big, so we should walk through
-        # it in one pass.
-        return list(filter(lambda checksum: checksum in checksums, self.all()))
+        """Check if the given checksums are stored in the remote.
+
+        There are two ways of performing this check:
+
+        - Traverse: Get a list of all the files in the remote
+            (traversing the cache directory) and compare it with
+            the given checksums.
+
+        - No traverse: For each given checksum, run the `exists`
+            method and filter the checksums that aren't on the remote.
+            This is done in parallel.
+
+        The reason for such an odd logic is that most of the remotes
+        take much shorter time to just retrieve everything they have under
+        a certain prefix (e.g. s3, gs, ssh, hdfs). Other remotes that can
+        check if particular file exists much quicker, use their own
+        implementation of cache_exists (see http, local).
+
+        Returns:
+            A list with checksums that were found in the remote
+        """
+
+        if self.no_traverse:
+            with ThreadPoolExecutor(max_workers=self.JOBS) as executor:
```

Comment:
question: for SSH the number of JOBS is pretty conservative because we are killing the network if the number is too high. What about this case? Could we safely use 100? 200?

Reply:
As you said, @shcheklein , it depends on the remote, for SSH we need to be careful, but I don't know if the rest of the remotes have a concern like that.

Reply:
@mroutis can we try to make it way higher? at least try 10x?

Reply:
just as a test first, to see the performance